core/logger/splunk.go

```diff
@@ -0,0 +1,186 @@
+package logger
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+)
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type event struct {
+	Time       int64           `json:"time"`                 // epoch time in nano-seconds
+	Host       string          `json:"host"`                 // hostname
+	Source     string          `json:"source,omitempty"`     // optional description of the source of the event; typically the app's name
+	SourceType string          `json:"sourcetype,omitempty"` // optional name of a Splunk parsing configuration; this is usually inferred by Splunk
+	Index      string          `json:"index,omitempty"`      // optional name of the Splunk index to store the event in; not required if the token has a default index set in Splunk
+	Event      json.RawMessage `json:"event"`                // zap json logger
+}
+
+type splunkSink struct {
+	HTTPClient    httpClient
+	nowFn         func() int64
+	URL           string // URL to Splunk HTTP Event Collector
+	Hostname      string // Hostname is the human-readable machine name sending logs
+	Token         string // Splunk HTTP Event Collector token
+	Source        string // Splunk source field value, description of the source of the event
+	SourceType    string // Splunk source type, optional name of a sourcetype field value
+	Index         string // Splunk Index, optional name of the Splunk index to store the event in
+	SkipTLSVerify bool   // Skip verifying the certificate of the HTTP Event Collector
+	events        [][]byte
+	newEvent      chan bool
+	sync.Mutex
+}
+
+func envVarOrDefault(key string, defaultValue string) string {
+	envVar := os.Getenv(key)
+	if envVar == "" {
+		return defaultValue
+	}
+	return envVar
+}
+
+func initialize(s *splunkSink) *splunkSink {
+	tr := &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: s.SkipTLSVerify}}
+	s.HTTPClient = &http.Client{Timeout: time.Second * 20, Transport: tr}
+
+	if s.Hostname == "" {
+		s.Hostname, _ = os.Hostname()
+	}
+	if s.nowFn == nil {
+		s.nowFn = func() int64 {
+			return time.Now().UnixNano()
+		}
+	}
+	s.newEvent = make(chan bool, 1)
+
+	runSync := func() {
+		err := s.Sync()
+		if err != nil {
+			log.Printf("Error sending to Splunk: %v\n", err)
+		}
+	}
+
+	go func() {
+		ticker := time.NewTicker(1 * time.Second)
+		atLeastOneEvent := false
+		for {
+			select {
+			case <-s.newEvent:
+				atLeastOneEvent = true
+				if len(s.events) > 50 {
+					atLeastOneEvent = false
+					go runSync()
```

Comment:
eep, feels like this could be unbounded. We probably would prefer a bounded channel that drops events if it gets full and one dedicated processing routine.